### PR TITLE
Add length checks to Array prototype functions

### PR DESCRIPTION
**Pull request by pmarkee**
Related issue: #124

Signed-off-by: Peter Marki marpeter@inf.u-szeged.hu

### CODE REVIEW

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 1]
```diff
@@ -560,6 +568,7 @@ static Value builtinArrayConcat(ExecutionState& state, Value thisValue, size_t a
                 n += len;
                 array->setThrowsException(state, ObjectPropertyName(state.context()->staticStrings().length), Value(n), array);
             } else {
+                CHECK_ARRAY_LENGTH(n);
```
Comment (clover2123) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: Would you check the ES2015 [spec](https://www.ecma-international.org/ecma-262/6.0/#sec-array.prototype.concat) once again? From the spec, it says that at this point

> If n**≥**2^53-1, throw a TypeError exception.

**Comment by pmarkee** [comment]
Thanks @clover2123 , I changed the macro so we can still use it in both places.

**Comment by clover2123** [comment]
@pmarkee 
BTW, does `CHECK_ARRAY_LENGTH` macro will be used in other places also? If not, how about just writing the code in place instead of using the macro? I think it looks less intuitive.

**Comment by pmarkee** [comment]
I checked that there are other functions where this check is also missing (`push`, `splice`, `unshift`), I will update the patch with those as well. I think it would be better to keep the macro, but I don't mind removing it.

**Comment by clover2123** [comment]
@pmarkee um.. one more thing please. There are some conflicts about array length between ES5.1 and ES6, so could you add TCs for ES6 and guard the new code paths with `ESCARGOT_ENABLE_ES2015`? Thanks.

**Comment by pmarkee** [comment]
Unfortunately it's quite hard to test this because of the very large array lengths. Let's say we want to trigger this error with a length of `2^54`. To do that, we need an array with a length of `2^27`, and each element also has to be an array with `2^27` length. (When spreading it out, the new length will be `2^27 * 2^27 = 2^54`) It requires a lot of memory to store this. I didn't see any specific test case for this in `js_vendor_tc` either.

**Comment by clover2123** [comment]
@pmarkee 
I see. I think we don't have to guard the new code path either because there are no conflicts within vendor tcs. 
BTW, Could you check the ES6 once more? I found that `push` and `slice` also check the array length as belows. So, the lengh check macro is necessary for these functions too.
- [Array.prototype.push](https://www.ecma-international.org/ecma-262/6.0/#sec-array.prototype.push)
> If len + argCount > 2^53-1, throw a TypeError exception.
- [Array.prototype.splice](https://www.ecma-international.org/ecma-262/6.0/#sec-array.prototype.splice)
> If len+insertCount−actualDeleteCount > 25^3-1, throw a TypeError exception.

**Comment by pmarkee** [comment]
I mentioned above that the es6 versions of `push()` and `splice()` are not implemented yet in escargot, only the es5 versions. I plan to update `Array.prototype` functions to their es6 versions soon (as discussed below #140 ), so on second thought maybe this PR isn't necessary at all, I could just add the length checks in that patch and close this PR. Would that be ok?

**Comment by yichoi** [comment]
@pmarkee please rebase it to the master.

**Comment by pmarkee** [comment]
@yichoi done, rebased to master.

**Pull request merged**